### PR TITLE
[Graphbolt] Enable cuda optimizations for UVAIndexSelect.

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description

This PR copies the kernel optimizations from https://github.com/dmlc/dgl/blob/master/src/array/cuda/uvm/array_index_select_uvm.cu.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/src/cuda/index_select_impl.cu [hunk 1]
```diff
@@ -46,19 +107,41 @@ torch::Tensor UVAIndexSelectImpl_(torch::Tensor input, torch::Tensor index) {
...
+        input_ptr, input_len, index_sorted_ptr, return_len, ret_ptr,
+        permutation_ptr);
+  } else {
+    dim3 block(512, 1);
+    while (static_cast<int64_t>(block.x) >= 2 * feature_size) {
+      block.x >>= 1;
+      block.y <<= 1;
+    }
+    const dim3 grid((return_len + block.y - 1) / block.y);
+    if (feature_size * sizeof(DType) < 2 * GPU_CACHE_LINE_SIZE) {
```
Comment (czkkkkkk) [inline graphbolt/src/cuda/index_select_impl.cu:133]: @mfbalin Could you explain why we choose `IndexSelectMultiKernel` in this case? From my understanding, `IndexSelectMultiKernelAligned` should be better than `IndexSelectMultiKernel` in any cases.

File: graphbolt/src/cuda/index_select_impl.cu [hunk 2]
```diff
@@ -46,19 +107,41 @@ torch::Tensor UVAIndexSelectImpl_(torch::Tensor input, torch::Tensor index) {
                                       .dtype(input.dtype())
                                       .device(c10::DeviceType::CUDA));
   DType* input_ptr = input.data_ptr<DType>();
-  IdType* index_ptr = index.data_ptr<IdType>();
   DType* ret_ptr = ret.data_ptr<DType>();
+
+  // Sort the index to improve the memory access pattern.
+  torch::Tensor sorted_index, permutation;
+  std::tie(sorted_index, permutation) = torch::sort(index);
```
Comment (mfbalin) [inline graphbolt/src/cuda/index_select_impl.cu:114]: Can we use CUB sort here? We can see the shape of the input, determine the number of bits required to radix sort the input, and use that for the sort. Should be faster. Like:
```
auto N = input.size(0);
int i = 0;
while ((1ll << i) < N)
  i++;
// pass i into cub::radixsort
```

**Review by mfbalin — CHANGES_REQUESTED** [review]
This code can be optimized further for future use cases. I am planning on revisiting it later.

File: graphbolt/src/cuda/index_select_impl.cu (hunk 1 above)
Comment (mfbalin) [reply graphbolt/src/cuda/index_select_impl.cu:133]: Using fewer SMs will be important when we want to overlap the copies with other ongoing computations. I will revisit this code to minimize SM usage further.

File: graphbolt/src/cuda/index_select_impl.cu [hunk 3]
```diff
@@ -46,19 +107,41 @@ torch::Tensor UVAIndexSelectImpl_(torch::Tensor input, torch::Tensor index) {
...
+  const int64_t* permutation_ptr = permutation.data_ptr<int64_t>();
+
   cudaStream_t stream = 0;
-  dim3 block(512, 1);
-  // Find the smallest block size that can fit the feature_size.
-  while (static_cast<int64_t>(block.x) >= 2 * feature_size) {
-    block.x >>= 1;
-    block.y <<= 1;
+
+  if (feature_size == 1) {
```
Comment (frozenbugs) [inline graphbolt/src/cuda/index_select_impl.cu:120]: Just to confirm, it the feature_size here the size of feature for each node or each edge?
Can you add documentation at the top of the cu file to explain how each option here improves each situation?

**Comment by frozenbugs** [comment]
Did you test the throughput of the new work compare to existing code?

File: graphbolt/src/cuda/index_select_impl.cu (hunk 2 above)
Comment (czkkkkkk) [reply graphbolt/src/cuda/index_select_impl.cu:114]: > I didn't author the if checks in this code. In my opinion, the threshold should be modified to simply
<= GPU_CACHE_LINE_SIZE. I think aligned is faster otherwise. You can benchmark it for different values to find out which one is better for sure. When there is 1 or less than 1 cache line of feature dimension, then the nonaligned version is more resource-efficient, meaning has the potential to use fewer SMs.

I see. I will only check the size with one GPU_CACHE_LINE_SIZE.

> Can we use CUB sort here?

Yes, we can. But it requires large changes. I will enable it in a later PR.

File: graphbolt/src/cuda/index_select_impl.cu (hunk 3 above)
Comment (czkkkkkk) [reply graphbolt/src/cuda/index_select_impl.cu:120]: > is the feature_size here the size of feature for each node or each edge?

Can be either node or edge feature. The index_select operator is designed for our feature store.

**Comment by czkkkkkk** [comment]
> Did you test the throughput of the new work compare to existing code?

Not yet.

File: tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py [hunk 4]
```diff
@@ -136,44 +138,34 @@ def test_torch_based_feature(in_memory):
...
+    )
+    tensor = torch.arange(0, reduce(mul, shape), dtype=dtype).reshape(shape)
+    test_tensor = tensor.clone().detach()
+    test_tensor_cuda = test_tensor.cuda()
+    pinned_tensor = tensor.pin_memory()
 
-    feature_a = gb.TorchBasedFeature(a)
-    feature_b = gb.TorchBasedFeature(b)
-    feature_c = gb.TorchBasedFeature(c)
+    feature = gb.TorchBasedFeature(pinned_tensor)
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py:154]: better to 
```
feature = gb.TorchBasedFeature(tensor).pin_memory_()
```
Instead of pin the tensor directly (better align with our API).

File: tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py [hunk 5]
```diff
@@ -136,44 +138,34 @@ def test_torch_based_feature(in_memory):
...
-        feature_b.read(torch.tensor([1], dtype=idtype).cuda()),
-        torch.tensor([[[4, 5], [6, 7]]], dtype=dtype).cuda(),
+        feature.read(torch.tensor([0], dtype=idtype).cuda()),
+        test_tensor_cuda[[0]],
     )
-    assert feature_b.read(torch.tensor([1], dtype=idtype).cuda()).is_cuda
+    assert feature.read(torch.tensor([0], dtype=idtype).cuda()).is_cuda
 
-    assert feature_c.read().is_cuda
+    assert feature.read().is_cuda
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py:164]: duplicate with L157.
```suggestion
```

File: tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py [hunk 6]
```diff
@@ -136,44 +138,34 @@ def test_torch_based_feature(in_memory):
...
+    test_tensor = tensor.clone().detach()
+    test_tensor_cuda = test_tensor.cuda()
+    pinned_tensor = tensor.pin_memory()
 
-    feature_a = gb.TorchBasedFeature(a)
-    feature_b = gb.TorchBasedFeature(b)
-    feature_c = gb.TorchBasedFeature(c)
+    feature = gb.TorchBasedFeature(pinned_tensor)
 
+    assert torch.equal(feature.read(), test_tensor_cuda)
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py:156]:
```suggestion
    # Test read entire pinned feature, the result should be on cuda.
    assert torch.equal(feature.read(), test_tensor_cuda)
```

File: tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py [hunk 7]
```diff
@@ -136,44 +138,34 @@ def test_torch_based_feature(in_memory):
...
-    )
-    assert feature_a.read(torch.tensor([0], dtype=idtype).cuda()).is_cuda
-    assert torch.equal(
-        feature_b.read(torch.tensor([1], dtype=idtype).cuda()),
-        torch.tensor([[[4, 5], [6, 7]]], dtype=dtype).cuda(),
+        feature.read(torch.tensor([0], dtype=idtype).cuda()),
+        test_tensor_cuda[[0]],
     )
-    assert feature_b.read(torch.tensor([1], dtype=idtype).cuda()).is_cuda
+    assert feature.read(torch.tensor([0], dtype=idtype).cuda()).is_cuda
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py:162]:
```suggestion

    # Test read pinned feature with idx on cuda, the result should be on cuda. 
    assert feature.read(torch.tensor([0], dtype=idtype).cuda()).is_cuda
```

File: tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py [hunk 8]
```diff
@@ -136,44 +138,34 @@ def test_torch_based_feature(in_memory):
...
 
-    assert feature_c.read().is_cuda
+    assert feature.read().is_cuda
     assert torch.equal(
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/impl/test_torch_based_feature_store.py:165]:
```suggestion
    # Test read pinned feature with idx on cpu, the result should be on cpu.
    assert torch.equal(
```

File: graphbolt/src/cuda/utils.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  while (ret > dim) {
+    ret = ret >> 1;
+  }
+  return ret;
+}
+
+}  // namespace cuda
+}  // namespace graphbolt
+
+#endif  // GRAPHBOLT_CUDA_UTILS_H_
```
Comment (frozenbugs) [inline graphbolt/src/cuda/utils.h:36]: no newline.

File: graphbolt/src/cuda/utils.h [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ * @brief CUDA utilities.
+ */
+
+#ifndef GRAPHBOLT_CUDA_UTILS_H_
+#define GRAPHBOLT_CUDA_UTILS_H_
+
+namespace graphbolt {
+namespace cuda {
+
+// The cache line size of GPU
```
Comment (frozenbugs) [inline graphbolt/src/cuda/utils.h:14]:
```suggestion
// The cache line size of GPU.
```

File: graphbolt/src/cuda/utils.h [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+#define GPU_CACHE_LINE_SIZE 128
+// The max number of threads per block
```
Comment (frozenbugs) [inline graphbolt/src/cuda/utils.h:16]:
```suggestion
// The max number of threads per block.
```

File: graphbolt/src/cuda/utils.h [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+#define CUDA_MAX_NUM_THREADS 1024
+
+/** @brief Calculate the number of threads needed given the dimension length.
```
Comment (frozenbugs) [inline graphbolt/src/cuda/utils.h:19]: What's "dimension length"? dimension or length? of length of the feature?

File: graphbolt/src/cuda/utils.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * It finds the biggest number that is smaller than min(dim, max_num_nthreads)
```
Comment (frozenbugs) [inline graphbolt/src/cuda/utils.h:21]: It finds the largest power of two that is less than or equal to the minimum of dim and max_num_threads.

File: graphbolt/src/cuda/utils.h [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+ * and is also power of two.
+ */
+inline int FindNumThreads(int dim, int max_num_threads = CUDA_MAX_NUM_THREADS) {
+  if (dim == 0) return 1;
+  int ret = max_num_threads;
+  while (ret > dim) {
+    ret = ret >> 1;
+  }
```
Comment (frozenbugs) [inline graphbolt/src/cuda/utils.h:29]: note: when max_num_threads is not power of two, the result is not guaranteed to be power of two, which conflicts with the documentation.
```suggestion
  int ret = 1;
  while ((ret << 1) <= std::min(dim, max_num_threads)) {
    ret <<= 1;
  }
```

Comment (frozenbugs) [reply graphbolt/src/cuda/utils.h:29]: consider rename dim to length.

File: graphbolt/src/cuda/utils.h (hunk 12 above)
Comment (czkkkkkk) [reply graphbolt/src/cuda/utils.h:19]: Should be "the size of the dimension to be processed". In the context of IndexSelectSingleKernel, it is the size of `index`.

File: graphbolt/src/cuda/utils.h (hunk 14 above)
Comment (czkkkkkk) [reply graphbolt/src/cuda/utils.h:29]: I used `size` to represent the size of a dimension.

**Review by frozenbugs — APPROVED** [review]

**Review by mfbalin — APPROVED** [review]
LGTM overall.

Noting for future revisit:
1) Sort can be changed to use fewer bits via cub sort
2) Regardless of the datatype, larger granularity accesses by each thread using a datatype such as float4 can be used to further reduce resource usage. In this case, the function will not depend on template parameter DType, but only on feature access granularity bytes. The types to use for copy are `uint8_t, uint16_t, uint32_t, float2, float4` and byte granularity being `1,2,4,8,16`. We will check if `gcd(feature dimension * sizeof DType, 16)`, and choose the appropriate type to perform the copy.
3) Limit the number of blocks used by taking a parameter indicating the max number of inflight accesses needed to saturate the PCI-e, as there will be diminishing returns because PCI-e bandwidth is limited.
